linen_examples/vae/train.py

```diff
@@ -0,0 +1,210 @@
+# limitations under the License.
+
+from absl import app
+from absl import flags
+
+import numpy as np
+import jax.numpy as jnp
+
+import jax
+from jax import random
+from jax.config import config
+config.enable_omnistaging()
+
+from flax import linen as nn
+from flax import optim
+
+import tensorflow as tf
+import tensorflow_datasets as tfds
+
+from utils import save_image
+
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_float(
+    'learning_rate', default=1e-3,
+    help=('The learning rate for the Adam optimizer.')
+)
+
+flags.DEFINE_integer(
+    'batch_size', default=128,
+    help=('Batch size for training.')
+)
+
+flags.DEFINE_integer(
+    'num_epochs', default=30,
+    help=('Number of training epochs.')
+)
+
+flags.DEFINE_integer(
+    'latents', default=20,
+    help=('Number of latent variables.')
+)
+
+
+class Encoder(nn.Module):
+  latents: int
+
+  @nn.compact
+  def __call__(self, x):
+    x = nn.Dense(500, name='fc1')(x)
+    x = nn.relu(x)
+    mean_x = nn.Dense(self.latents, name='fc2_mean')(x)
+    logvar_x = nn.Dense(self.latents, name='fc2_logvar')(x)
+    return mean_x, logvar_x
+
+
+class Decoder(nn.Module):
+
+  @nn.compact
+  def __call__(self, z):
+    z = nn.Dense(500, name='fc1')(z)
+    z = nn.relu(z)
+    z = nn.Dense(784, name='fc2')(z)
+    return z
+
+
+class VAE(nn.Module):
+  latents: int = 20
+
+  def setup(self):
+    self.encoder = Encoder(self.latents)
+    self.decoder = Decoder()
+
+  def __call__(self, x, z_rng):
+    mean, logvar = self.encoder(x)
+    z = reparameterize(z_rng, mean, logvar)
+    recon_x = self.decoder(z)
+    return recon_x, mean, logvar
+
+  def generate(self, z):
+    return nn.sigmoid(self.decoder(z))
+
+
+def reparameterize(rng, mean, logvar):
+  std = jnp.exp(0.5 * logvar)
+  eps = random.normal(rng, logvar.shape)
+  return mean + eps * std
+
+
+@jax.vmap
+def kl_divergence(mean, logvar):
+  return -0.5 * jnp.sum(1 + logvar - jnp.square(mean) - jnp.exp(logvar))
+
+
+@jax.vmap
+def binary_cross_entropy_with_logits(logits, labels):
+  logits = nn.log_sigmoid(logits)
+  return -jnp.sum(labels * logits + (1. - labels) * jnp.log(-jnp.expm1(logits)))
+
+
+def compute_metrics(recon_x, x, mean, logvar):
+  bce_loss = binary_cross_entropy_with_logits(recon_x, x).mean()
+  kld_loss = kl_divergence(mean, logvar).mean()
+  return {
+      'bce': bce_loss,
+      'kld': kld_loss,
+      'loss': bce_loss + kld_loss
+  }
+
+
+def model():
+  return VAE(latents=FLAGS.latents)
+
+
+@jax.jit
+def train_step(optimizer, batch, z_rng):
+  def loss_fn(params):
+    recon_x, mean, logvar = model().apply({'param': params}, batch, z_rng)
+
+    bce_loss = binary_cross_entropy_with_logits(recon_x, batch).mean()
+    kld_loss = kl_divergence(mean, logvar).mean()
+    loss = bce_loss + kld_loss
+    return loss, recon_x
+  grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+  _, grad = grad_fn(optimizer.target)
+  optimizer = optimizer.apply_gradient(grad)
+  return optimizer
+
+
+# @jax.jit
```

Comment:
Is this jit intentionally commented out? The original code had it.

Reply:
Sorry about that! I put it back in.